### PR TITLE
Add support for feature detection on FreeBSD/aarch64 (WIP)

### PR DESCRIPTION
**Pull request by myfreeweb**
On AArch64, FreeBSD does not currently provide `HWCAP` in the ELF auxiliary vector.
Reads of the ID registers are trapped though, and they are accessible in userspace.

- The register reading code could work on other OSes and even bare metal, so it's separated out into its own module.
- Seems like it could also be used on Linux.
- ~~TODO: The OS-dependent code should trap `SIGILL` before calling the reading function, in order to work on older OS versions that didn't expose the registers.~~  I guess it's fine™ to expect that to work.
- ~~TODO: The OS-dependent code should run the reading function on all cores (via thread affinity) and take a union of features, because there are [weird SoCs where cores have unequal features](https://medium.com/@jadr2ddude/a-big-little-problem-a-tale-of-big-little-gone-wrong-e7778ce744bb).~~ [On FreeBSD, the register view is the same on each core](https://reviews.freebsd.org/D17137#393947)

---
References:

- [Linux kernel docs: ARM64 CPU Feature Registers](https://www.kernel.org/doc/Documentation/arm64/cpu-feature-registers.txt)
- [FreeBSD r337739: Support reading from the arm64 ID registers from userspace.](https://reviews.freebsd.org/rS337739)
- [FreeBSD D17137: arm64: Add HWCAP support](https://reviews.freebsd.org/D17137)

### CODE REVIEW

**Comment by lu-zero** [comment]
How old are we talking regarding old oses? Trapping signals is _quite_ bad for libraries.

**Comment by myfreeweb** [comment]
@lu-zero well, older than FreeBSD 12 (which is currently in release candidate phase). The number of people who use Rust + FreeBSD + AArch64 is currently quite small, especially considering that rustc currently doesn't build on FreeBSD/aarch64 without [a very patch that hasn't landed yet](https://reviews.freebsd.org/D18417) :) So it should be fine to ignore that.

File: stdsimd/arch/detect/os/aarch64.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        unsafe { asm!("mrs $0, S3_0_C0_C6_0" : "=r"(aa64isar0)); }
+
+        let aes = bits_shift(aa64isar0, 7, 4) >= 1;
+        let pmull = bits_shift(aa64isar0, 7, 4) >= 2;
+        let sha1 = bits_shift(aa64isar0, 11, 8) >= 1;
+        let sha2 = bits_shift(aa64isar0, 15, 12) >= 1;
+        enable_feature(Feature::pmull, pmull);
+        // Crypto is specified as AES + PMULL + SHA1 + SHA2 per LLVM/hosts.cpp
+        enable_feature(Feature::crypto, aes && pmull && sha1 && sha2);
+        enable_feature(Feature::lse, bits_shift(aa64isar0, 23, 20) >= 1);
```
Comment (alexcrichton) [inline stdsimd/arch/detect/os/aarch64.rs]: According to [this page](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.100458_0200_00_en/lau1442496521966.html) the bits here correspond to the "atomic" feature, although I'm pretty unfamiliar with AArch64 feature names so `lse` may actually be that! Just wanted to confirm

File: stdsimd/arch/detect/os/aarch64.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        enable_feature(Feature::crc, bits_shift(aa64isar0, 19, 16) >= 1);
```
Comment (alexcrichton) [inline stdsimd/arch/detect/os/aarch64.rs]: This says >= 1 but [these docs](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.100458_0200_00_en/lau1442496521966.html) indicate the value may need to be exactly one?

File: stdsimd/arch/detect/os/aarch64.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                value.set(f as u32);
+            }
+        };
+
+        // ID_AA64ISAR0_EL1 - Instruction Set Attribute Register 0
+        let aa64isar0: u64;
+        unsafe { asm!("mrs $0, S3_0_C0_C6_0" : "=r"(aa64isar0)); }
+
+        let aes = bits_shift(aa64isar0, 7, 4) >= 1;
+        let pmull = bits_shift(aa64isar0, 7, 4) >= 2;
```
Comment (alexcrichton) [inline stdsimd/arch/detect/os/aarch64.rs]: According to [arm docs](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.100458_0200_00_en/lau1442496521966.html) should this only check for 0x2? 

Additionally, I wonder if this plus `crypto` below means that we should split up the `crypto` feature if AArch64 actually has multiple features here... (not necessary to do in this PR)

File: stdsimd/arch/detect/os/aarch64.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        let fphp = bits_shift(aa64pfr0, 19, 16) >= 1;
+        let asimd = bits_shift(aa64pfr0, 23, 20) < 0xF;
+        let asimdhp = bits_shift(aa64pfr0, 23, 20) >= 1;
+        enable_feature(Feature::fp, fp);
+        enable_feature(Feature::fp16, fphp);
+        // SIMD support requires float support - if half-floats are
+        // supported, it also requires half-float support:
+        enable_feature(Feature::asimd, fp && asimd && (!fphp | asimdhp));
+        // SIMD extensions require SIMD support:
+        enable_feature(Feature::rdm, asimd && bits_shift(aa64isar0, 31, 28) >= 1);
```
Comment (alexcrichton) [inline stdsimd/arch/detect/os/aarch64.rs]: Hm on [these docs](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.100403_0200_00_en/lau1443437525908.html) this is called  the RAS feature...

**Comment by gnzlbg** [comment]
I’d like to review this with calm tomorrow afternoon if it isn’t very
pressing.

w.r.t. signal-handling-based detection, there were some intrinsic issues
with this (e.g. how reliable is it that the proper exception will be raised
on a CPU that “doesn’t know” the instruction being executed?). However, in
some OSes the alternatives aren’t clear (e.g. windows), as long as “it
works for you” and doesn’t break anything I think it is something worth
getting experience with.

Would it be possible to get CI set up for aarch64-FreeBSD?
-- 
Dipl.-Ing. Gonzalo Brito Gadeschi
Institute of Aerodynamics and Chair of Fluid Mechanics
RWTH Aachen University
Wuellnerstraße 5a
D-52062 Aachen
Germany
Phone: ++49-(0)241-80-94821
Fax: ++49-(0)241-80-92257
E-mail:  g.brito@aia.rwth-aachen.de
Internet: www.aia.rwth-aachen.de

File: stdsimd/arch/detect/os/aarch64.rs (hunk 1 above)
Comment (myfreeweb) [reply stdsimd/arch/detect/os/aarch64.rs]: `lse` is `atomic`, look at the comments in the `Feature` struct and the Linux implementation

File: stdsimd/arch/detect/os/aarch64.rs (hunk 2 above)
Comment (myfreeweb) [reply stdsimd/arch/detect/os/aarch64.rs]: With these registers, the semantics is: higher number == higher level of support of this feature (which includes all previous ones). So e.g. if there's a next level of CRC32 implemented in the future, that CPU will have this value as 2.

e.g. Zircon implements checks as `>=`: https://fuchsia.googlesource.com/zircon/+/master/kernel/arch/arm64/feature.cpp#204

File: stdsimd/arch/detect/os/aarch64.rs (hunk 3 above)
Comment (myfreeweb) [reply stdsimd/arch/detect/os/aarch64.rs]: (see above)

This is a better example actually: this is the same bits of the same register. `>= 1` means AES is supported, `>= 2` means PMULL is supported.

I like the idea of splitting `crypto`. Looks like the idea here was to align to LLVM's target features, but it's okay to e.g. check for AES only when writing an AES function that would use the `crypto` target feature

**Comment by myfreeweb** [comment]
> I’d like to review this with calm tomorrow afternoon

Of course :)

> a CPU that “doesn’t know” the instruction being executed

Every aarch64 CPU knows `msr` with these arguments, they're just privileged instructions, and most (?) OSes trap them to allow them to succeed in userspace.

I can see how trying to detect the SIMD/etc instructions themselves via signal handling can be unreliable, yes.

>  Would it be possible to get CI set up for aarch64-FreeBSD?

I have a public buildbot..

File: stdsimd/arch/detect/os/aarch64.rs (hunk 4 above)
Comment (myfreeweb) [reply stdsimd/arch/detect/os/aarch64.rs]: That link is to `PFR0`, not `ISAR0`

**Comment by myfreeweb** [comment]
Update: no need to read the values on different cores, the kernel already presents the same view of the registers on all cores https://reviews.freebsd.org/D17137#393947 So this is good as is, pretty much. I can add some code comments about this I guess.

File: stdsimd/arch/detect/os/aarch64.rs (hunk 2 above)
Comment (alexcrichton) [reply stdsimd/arch/detect/os/aarch64.rs]: Ah ok, good to know! Mind adding a comment somewhere in this file about that? (doesn't have to be on every comparison)

File: stdsimd/arch/detect/os/aarch64.rs (hunk 1 above)
Comment (alexcrichton) [reply stdsimd/arch/detect/os/aarch64.rs]: Ok cool, thanks for the info!

File: stdsimd/arch/detect/os/aarch64.rs (hunk 3 above)
Comment (alexcrichton) [reply stdsimd/arch/detect/os/aarch64.rs]: Ok cool, let's move [discussion for `crypto` to an issue](https://github.com/rust-lang-nursery/stdsimd/issues/614)

File: stdsimd/arch/detect/os/aarch64.rs (hunk 4 above)
Comment (alexcrichton) [reply stdsimd/arch/detect/os/aarch64.rs]: Oh sorry I got mixed up with the blocks here, makes sense

**Comment by alexcrichton** [comment]
FWIW I personally think that we shouldn't try to trap instructions here, if FreeBSD handles this and "makes the instruction work" that's fine by me and if we're not compatible with older versions that's probably ok for now too

File: stdsimd/arch/detect/mod.rs [hunk 5]
```diff
@@ -63,6 +63,13 @@ pub use self::arch::Feature;
 mod bit;
 mod cache;
 
+cfg_if! {
+    if #[cfg(all(target_arch = "aarch64", any(target_os = "freebsd")))] {
```
Comment (gnzlbg) [inline stdsimd/arch/detect/mod.rs]: Why is the `any(target_os = "freebsd")` needed here?  (shouldn't it just be `target_os = "freebsd"` ?)

File: stdsimd/arch/detect/os/aarch64.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        let mut enable_feature = |f, enable| {
+            if enable {
+                value.set(f as u32);
+            }
+        };
+
+        // ID_AA64ISAR0_EL1 - Instruction Set Attribute Register 0
+        let aa64isar0: u64;
+        unsafe { asm!("mrs $0, S3_0_C0_C6_0" : "=r"(aa64isar0)); }
```
Comment (gnzlbg) [inline stdsimd/arch/detect/os/aarch64.rs]: @japaric do we have an intrinsic for this or is one planned ?

File: stdsimd/arch/detect/os/aarch64.rs (hunk 3 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/os/aarch64.rs]: This looks resolved to me (splitting `crypto` is an orthogonal issue).

File: stdsimd/arch/detect/os/aarch64.rs (hunk 2 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/os/aarch64.rs]: It would be good if the top level comments could contain links to zircon and every other place that was relevant when coming up with the implementation.

File: stdsimd/arch/detect/os/aarch64.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        // supported, it also requires half-float support:
+        enable_feature(Feature::asimd, fp && asimd && (!fphp | asimdhp));
+        // SIMD extensions require SIMD support:
+        enable_feature(Feature::rdm, asimd && bits_shift(aa64isar0, 31, 28) >= 1);
+        enable_feature(Feature::dotprod, asimd && bits_shift(aa64isar0, 47, 44) >= 1);
+        enable_feature(Feature::sve, asimd && bits_shift(aa64pfr0, 35, 32) >= 1);
+
+        // ID_AA64ISAR1_EL1 - Instruction Set Attribute Register 1
+        let aa64isar1: u64;
+        unsafe { asm!("mrs $0, S3_0_C0_C6_1" : "=r"(aa64isar1)); }
```
Comment (gnzlbg) [inline stdsimd/arch/detect/os/aarch64.rs]: What's the "reference" for the contents of `msr` ? That is, how should it be queried, what do the bits in the output registers mean, etc. ? It might be worth it to link to the reference in the top-level comment of the module and minimally explain how this works (not that it is hard to follow, but it'll make lives easier of those having to update this in the future).

File: stdsimd/arch/detect/os/aarch64.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
+//! Run-time feature detection for Aarch64 on any OS that traps the mrs instruction.
```
Comment (parched) [inline stdsimd/arch/detect/os/aarch64.rs]: Perhaps this file could go coresimd as it would be useful for bare metal too?

File: stdsimd/arch/detect/os/aarch64.rs (hunk 6 above)
Comment (parched) [reply stdsimd/arch/detect/os/aarch64.rs]: @gnzlbg That's the `rsr64` intrinsic added in the other open PR but it doesn't this form yet.

@myfreeweb Probably better to use `ID_AA64ISAR0_EL1` instead of  `S3_0_C0_C6_0` in the `asm!` to make it more readable/easier to verify.

File: stdsimd/arch/detect/os/aarch64.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+        let fp = bits_shift(aa64pfr0, 19, 16) < 0xF;
+        let fphp = bits_shift(aa64pfr0, 19, 16) >= 1;
+        let asimd = bits_shift(aa64pfr0, 23, 20) < 0xF;
+        let asimdhp = bits_shift(aa64pfr0, 23, 20) >= 1;
+        enable_feature(Feature::fp, fp);
+        enable_feature(Feature::fp16, fphp);
+        // SIMD support requires float support - if half-floats are
+        // supported, it also requires half-float support:
+        enable_feature(Feature::asimd, fp && asimd && (!fphp | asimdhp));
```
Comment (parched) [inline stdsimd/arch/detect/os/aarch64.rs]: I think it would be better to just have `asimd` as the second arg here. Then add an assert for `fp == asimd` and `fphp == asimdhp` as these are required by the architecture.

**Comment by gnzlbg** [comment]
> The register reading code could work on other OSes and even bare metal, so it's separated out into its own module.

So I think this is a good idea. We do the same for `x86` where we have it as its own os, I think it is worth it to do this for `aarch64` as well.  Since this is part of `libstd` , it is unlikely that people will use it directly on bare metal in the near future, but there is some interest to do this with `x86` as well and there are some ways to do that (e.g. split those into different `no_std` crates that people can include without pulling in std, portability check, etc.).

> Seems like it could also be used on Linux.

IIRC `msr` was a privileged instruction on Linux, but maybe I remember wrong?

> TODO: The OS-dependent code should run the reading function on all cores (via thread affinity) and take a union of features, because there are weird SoCs where cores have unequal features. On FreeBSD, the register view is the same on each core

`stdsimd` is used in `libstd`, so it requires OS support, we do have some workarounds over some specific known CPU bugs, but we don't have any workarounds that assume an "adversary" OS that tries to break your process. 

On Linux, we ask the OS "What is this process allowed to do?" via the OS interfaces (e.g. auxiliary vectors). If Linux decides to migrate the process to a different core, it has to make sure that the core is at least as capable as the one we are running on because we can cache the answer to the question. If the FreeBSD way of querying this information is via the `msr` register, then they have to make sure about that too. 

When the OS fails, all bets are off anyways. For example, see [A tale of big.LITTLE gone wrong](https://medium.com/@jadr2ddude/a-big-little-problem-a-tale-of-big-little-gone-wrong-e7778ce744bb), where Samsung Exynos aarch64 LITTLE cores had atomic support but the big cores did not (usually big cores are "better" than little cores, and the OS assumed that all SoCs were like this).  So the Apps would start in LITTLE cores, do run-time feature detection, detect that they can use atomics, cache that, start using them, and when some time the OS moved them to a big core, then they would SIGILL because the big cores did not support atomics. Working around "bugs" like these here feels like trying to do too much: we'd need OS APIs that allow us to query the set of potential cores the process might be moved to, and APIs to query the features of these cores.

> TODO: The OS-dependent code should trap SIGILL before calling the reading function, in order to work on older OS versions that didn't expose the registers. I guess it's fine™ to expect that to work.

What's the recommended FreeBSD way of handling this ? We don't support Windows+aarch64 but the recommended approach to do run-time feature detection there is to "try to use an intrinsic, and use SEH to handle the error if its not available" (see https://github.com/rust-lang-nursery/stdsimd/issues/120#issuecomment-339818841) . I agree with @lu-zero and @alexcrichton that we should avoid this if possible.

File: stdsimd/arch/detect/os/aarch64.rs (hunk 8 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/os/aarch64.rs]: @parched we don't really have a mechanism to expose this from `core::arch` yet (and it is unclear how to do that). I'd leave doing that to "future work".

File: stdsimd/arch/detect/os/aarch64.rs (hunk 9 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/os/aarch64.rs]: This is how we do it on Linux: https://github.com/rust-lang-nursery/stdsimd/blob/master/stdsimd/arch/detect/os/linux/aarch64.rs#L148

```rust
            let asimd = self.fp && self.asimd && (!self.fphp | self.asimdhp);
            enable_feature(Feature::asimd, asimd);
```

Maybe it would be worth it to somehow keep both implementations in sync.

Comment (parched) [reply stdsimd/arch/detect/os/aarch64.rs]: Ah ok hadn't seen that, yeah, best to keep them in sync.

**Comment by parched** [comment]
>> Seems like it could also be used on Linux.
>>
>IIRC msr was a privileged instruction on Linux, but maybe I remember wrong?

Yes this works on Linux now too, since [version 4.11](https://github.com/torvalds/linux/commit/4aa8a472c33f4024381ade56bef008a9c566366a#diff-9764502b1d5418a9c00b3b235fd51bdc). 

Whether MRS is privileged depends on what register is being read, i.e., only ones ending with '_EL0' can possibly be read directly from userspace. EL1 ones like these cause an exception and must be emulated by the kernel if it wants (normally it will abort the process with SIGILL or similar).

**Comment by gnzlbg** [comment]
> Yes this works on Linux now too, since version 4.11.

Cool! We should explore using this in Linux then, it is way easier than using auxiliary vectors.

File: stdsimd/arch/detect/mod.rs (hunk 5 above)
Comment (myfreeweb) [reply stdsimd/arch/detect/mod.rs]: A subtle hint for potential future reuse of this module by other OSes :) I guess a comment would be better.

**Comment by myfreeweb** [comment]
> If the FreeBSD way of querying this information is via the msr register, then they have to make sure about that too

Yep, the emulated msr only returns features supported by all cores. (Looks like you replied before I updated the original post.)

> see A tale of big.LITTLE gone wrong

That's where I linked with the "weird SoCs where cores have unequal features" text :)

> What's the recommended FreeBSD way of handling this ?

Similar to Windows, I guess. For an API that's literally "run a privileged instruction and expect it to be emulated", not much you can do other than trapping SIGILL.

But we can just ignore this. The FreeBSD/aarch64 userbase is small enough that "just upgrade your OS to >=12" is fine :)

File: stdsimd/arch/detect/mod.rs (hunk 5 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/mod.rs]: Note that because we use `#[path]` here any os wanting to use the `aarch64` module can just use it by including it directly below - so there is no need to split the per-OS logic into two `cfg_if!`s.

File: stdsimd/arch/detect/os/aarch64.rs (hunk 6 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/os/aarch64.rs]: @myfreeweb see @parched suggestion to use `ID_AA64ISAR0_EL1` instead of `S3_0_C0_C6_0`.

File: stdsimd/arch/detect/os/aarch64.rs (hunk 2 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/os/aarch64.rs]: You have some links in the PR message, please consider adding all that information to the top of the module to avoid people from having to find the commit / PR to get it.

File: stdsimd/arch/detect/os/aarch64.rs (hunk 7 above)
Comment (gnzlbg) [reply stdsimd/arch/detect/os/aarch64.rs]: The links are in the PR, and there is already another comment suggesting adding these links to the module, so I am marking this as resolved.

**Comment by gnzlbg** [comment]
> But we can just ignore this. The FreeBSD/aarch64 userbase is small enough that "just upgrade your OS to >=12" is fine :)

We can leave adding support for FreeBSD == 11 to "future work". If someone complains, I'd guess we can just disable this with a config flag or by querying the FreeBSD version somehow later. 

This is looking good, just a couple of nits open (don't forget to run `cargo fmt --all` since that is the only CI failure)!

**Comment by myfreeweb** [comment]
Updated with comments and suggested `cfg`.

> cargo fmt --all

The CI rustfmt failure is:

```
error: component 'rustfmt' for target 'x86_64-unknown-linux-gnu' is unavailable for download
```

**Comment by gnzlbg** [comment]
Thank you a lot @myfreeweb !
